### ISSUE TITLE
Running tests in Electron main process

### ISSUE THREAD

**Issue by boyangwang** [facebook-atom/jest-electron-runner#18]
If I want to test a piece of code that's meant to run in Electron main process, is there anyway to configure jest-electron-runner to achieve that?

Currently `require('electron').ipcRenderer` can be imported and used, but `require('electron').ipcMain` is undefined.

Using `require('electron').remote.ipcMain` is not ideal because that would change the code to be tested - in actual execution it will be in main process

**Comment by boujeepossum** [comment]
it should be possible to do that, but it'll require some extra code in the runner.
it'll also probably have to be a separate runner.
e.g.
```
runner: "@jest-runner/electron/main-process"
```

i'll see if i have time to make it work!

**Comment by boyangwang** [comment]
@aaronabramov thanks for your reply. If you can roughly explain your idea and impl plan, I might be able to do a PR. Otherwise, there's no hurry

**Comment by boujeepossum** [comment]
@boyangwang that would be much appreciated!
here's what it looks like right now:

```
+-----------+    +------------+    +-------------+ 
|           |RPC |            |RPC |  render proc| 
| jest      | -->| main proc  |--->|+-----------+| 
|           |    |            |    ||           || 
+-----------+    +------------+    ||jest-runner|| 
                                   |+-----------+| 
                                   +-------------+ 
```
jest spawns an electron, then interacts with the main process, which in turn creates browser window where it runs test using `jest-runner`
what needs to be done is:
1. make the existing runner configurable (or subclass it?)
2. make another RPC module that will run tests inside main process (instead of creating windows) see: https://github.com/facebook-atom/jest-electron-runner/blob/master/packages/electron/src/rpc/JestWorkerRPC.js
3. create a module that can run tests in main process (similar to https://github.com/facebook-atom/jest-electron-runner/blob/master/packages/electron/src/browser_window_injected_code.js)

I expect there to be a lot of unexpected complexity though. for example right now browser windows are created and dropped for every test, so there's prefect isolation. with main process we'll probably have to spawn electron every single time (which will probably be much slower)

**Comment by JM-Mendez** [comment]
@boyangwang I submitted pr https://github.com/facebook-atom/jest-electron-runner/pull/19 that runs the tests in a node process, rather than in a renderer process. I did it for different reasons, but I think it should work. Although after seeing this issue, I see I may have to isolate the tests better.  But I'll wait to see what @aaronabramov says first before changing anything else.

**Comment by boujeepossum** [comment]
published [v0.2.2](https://www.npmjs.com/package/@jest-runner/electron)
